### PR TITLE
NV15: Speed up the migration

### PR DESCRIPTION
**Pull request by arajasek**
The special Sectors migrating logic is as follows:

- For each miner, if the Head CID is found in the cache, use the cached value (same as any other actor)
- If not load the current state, do the quick steps like migrating MinerInfo
- Check the cache for the sectors AMT root -- if found, use the cached value and return back to miner migration
  - Note that this step will cover many hundreds of thousands of "empty" miners in the state tree
- If not found, check the cache to see if a previous sectors AMT has been cached **for this miner**. If not found, manually migrate all sectors, and insert the incoming and outgoing AMTs into the cache (specific to the miner)
- If the cache does have a previous sectors AMT cached for this miner, load it and its corresponding migrated AMT. Diff the previous AMT against the current one, mutating the previous migrated AMT appropriately. Update the cached AMT roots.

### CODE REVIEW

File: actors/builtin/miner/partition_state.go [hunk 1]
```diff
@@ -56,10 +58,7 @@ const PartitionExpirationAmtBitwidth = 4
 const PartitionEarlyTerminationArrayAmtBitwidth = 3
 
 // Value type for a pair of raw and QA power.
-type PowerPair struct {
```
Comment (arajasek) [inline actors/builtin/miner/partition_state.go:59]: Unrelated to this migration, sorry.

File: actors/migration/nv15/top.go [hunk 2]
```diff
@@ -57,6 +53,20 @@ type MigrationCache interface {
 	Load(key string, loadFunc func() (cid.Cid, error)) (cid.Cid, error)
 }
 
+func CidKey(c cid.Cid) string {
```
Comment (arajasek) [inline actors/migration/nv15/top.go:56]: I'm just using one catch-all key for all migrated objects (unlike the ActorHeadKey etc we had before). Is that bad?

File: actors/migration/nv15/top.go [hunk 3]
```diff
@@ -226,10 +238,9 @@ func MigrateStateTree(ctx context.Context, store cbor.IpldStore, actorsRootIn ci
 
 type actorMigrationInput struct {
 	address    address.Address // actor's address
-	balance    abi.TokenAmount // actor's balance
-	head       cid.Cid         // actor's state head CID
-	priorEpoch abi.ChainEpoch  // epoch of last state transition prior to migration
-	cache      MigrationCache  // cache of existing cid -> cid migrations for this actor
+	head       cid.Cid
```
Comment (arajasek) [inline actors/migration/nv15/top.go:241]: balance shouldn't have been in here, i don't think

File: actors/migration/nv15/top.go [hunk 4]
```diff
@@ -246,7 +257,7 @@ type actorMigration interface {
 
 type migrationJob struct {
 	address.Address
-	states6.Actor
+	Actor states6.Actor
```
Comment (arajasek) [inline actors/migration/nv15/top.go:260]:
```suggestion
	states6.Actor
```
Stale, sorry.

File: actors/migration/nv15/top.go [hunk 5]
```diff
@@ -57,6 +53,20 @@ type MigrationCache interface {
 	Load(key string, loadFunc func() (cid.Cid, error)) (cid.Cid, error)
 }
 
+func CidKey(c cid.Cid) string {
+	return c.String()
+}
+
+func MinerPrevSectorsInKey(m address.Address) string {
```
Comment (arajasek) [inline actors/migration/nv15/top.go:60]: we'll likely wanna persist these to allow for restarts after premigrations have run. maybe.

File: actors/migration/nv15/miner.go [hunk 6]
```diff
@@ -19,63 +25,184 @@ type minerMigrator struct{}
...
+func migrateSectors(ctx context.Context, store adt.Store, cache MigrationCache, minerAddr address.Address, inRoot cid.Cid) (cid.Cid, error) {
+	return cache.Load(CidKey(inRoot), func() (cid.Cid, error) {
+		inArray, err := adt.AsArray(store, inRoot, miner6.SectorsAmtBitwidth)
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to read sectors array: %w", err)
+		}
+
+		okIn, prevInRoot, err := cache.Read(MinerPrevSectorsInKey(minerAddr))
+		if err != nil {
+			// TODO: Should we instead absorb errors like these and proceed cache-lessly?
```
Comment (arajasek) [inline actors/migration/nv15/miner.go:72]: Opinions solicited

File: actors/migration/nv15/miner.go [hunk 7]
```diff
@@ -19,63 +25,184 @@ type minerMigrator struct{}
...
 
 func (m minerMigrator) migratedCodeCID() cid.Cid {
 	return builtin7.StorageMinerActorCodeID
 }
 
-func migrateSectors(ctx context.Context, store cbor.IpldStore, inRoot cid.Cid) (cid.Cid, error) {
-	ctxStore := adt.WrapStore(ctx, store)
-	inArray, err := adt.AsArray(ctxStore, inRoot, miner6.SectorsAmtBitwidth)
+func migrateSectors(ctx context.Context, store adt.Store, cache MigrationCache, minerAddr address.Address, inRoot cid.Cid) (cid.Cid, error) {
+	return cache.Load(CidKey(inRoot), func() (cid.Cid, error) {
```
Comment (arajasek) [inline actors/migration/nv15/miner.go:64]: The overwhelming majority of actors should be covered in the cache here:

- Even if this is the first premigration, the majority of miners are "empty" and so have identical sectors AMT roots
- If a premigration HAS already run, the non-empty miners will very likely be covered here because the sectors AMT doesn't change that much -- new sectors and extensions are the only things that can change it, i think (._..and CompactPartitions_)

Comment (ZenGround0) [reply actors/migration/nv15/miner.go:64]: For peace of mind use the pattern we established last year to type cache entries: 
1. Types: https://github.com/filecoin-project/specs-actors/blob/a2aec9c6af0e55ff9f7da257b9b5ca4c3e88b8e9/actors/migration/nv10/top.go#L47
2. Usage: https://github.com/filecoin-project/specs-actors/blob/release/v3/actors/migration/nv10/miner.go#L40

Data collisions (the thing being hashed as an AMT somehow has the same bytes as the thing being hashed as miner state) really should be impossible but it would be nice not to think about this at all.

File: actors/migration/nv15/top.go (hunk 2 above)
Comment (ZenGround0) [reply actors/migration/nv15/top.go:56]: Ah nice missed this call out.  Its probably fine but better to be safe.  More details in [this review comment](https://github.com/filecoin-project/specs-actors/pull/1548#discussion_r777055395)

File: actors/migration/nv15/top.go (hunk 5 above)
Comment (ZenGround0) [reply actors/migration/nv15/top.go:60]: This is a lotus side thing but I agree

File: gen/gen.go [hunk 8]
```diff
@@ -176,7 +176,7 @@ func main() {
 		miner.Deadline{},
 		miner.Partition{},
 		miner.ExpirationSet{},
-		miner.PowerPair{},
+		//miner.PowerPair{}, // Aliased
```
Comment (ZenGround0) [inline gen/gen.go:179]: nit: from v6

File: actors/migration/nv15/miner.go (hunk 6 above)
Comment (ZenGround0) [reply actors/migration/nv15/miner.go:72]: Either way is fine but I think we can rely on cache not erroring on read of a valid key.  If the cache is written right I expect the errors to be local system dependent only.

File: actors/migration/nv15/miner.go [hunk 9]
```diff
@@ -19,63 +25,184 @@ type minerMigrator struct{}
...
+				return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+			}
+
+			for _, change := range diffs {
+				switch change.Type {
+				case amt.Remove:
+					if err := prevOutSectors.Delete(change.Key); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+					}
+				case amt.Add:
```
Comment (ZenGround0) [inline actors/migration/nv15/miner.go:123]: I think this needs to `fallthrough` otherwise you don't migrate sectors added to the amt in between.  https://go.dev/play/p/fPxSEXpevf7. This highlights the need for thorough testing.  A good place to draw inspiration is the deal label migration test: https://github.com/filecoin-project/specs-actors/pull/1551/files#diff-5f4089c6e3f8495e10bfa13e18ff8f0e34cb787c56256e0bfabc576f0253f66aR36

**Review by ZenGround0 — CHANGES_REQUESTED** [review]
Other than the one bug this looks good.  It will be important to thoroughly test edge cases and the actors vm test framework works pretty well for this.

File: actors/migration/nv15/miner.go [hunk 10]
```diff
@@ -19,63 +25,184 @@ type minerMigrator struct{}
...
+	outState := fromv6State(inState)
+	ctxStore := adt.WrapStore(ctx, store)
+
+	sectorsOut, err := migrateSectors(ctx, ctxStore, in.cache, in.address, inState.Sectors)
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("failed to migrate sectors: %w", err)
 	}
 
-	outState := miner7.State{
```
[the remark this answers is not in the document]
Comment (laudiacay) [reply actors/migration/nv15/miner.go:30]: honestly i found the old pattern hard to read- it was hard to see which fields were changing and which weren't? but I see this concern. valid.

**Review by laudiacay — CHANGES_REQUESTED** [review]
I don't have any comments outside of what zen said other than "it would be cool to have a test for this". 

looks good and correct unless my lying eyes deceive me, and tests are good for catching code reviewers' lying eyes :P agree that the deal label migration code would be a good start

File: actors/builtin/miner/partition_state.go (hunk 1 above)
Comment (arajasek) [reply actors/builtin/miner/partition_state.go:59]: Ack

File: actors/migration/nv15/top.go (hunk 2 above)
Comment (arajasek) [reply actors/migration/nv15/top.go:56]: Ack, will do!

File: actors/migration/nv15/miner.go (hunk 7 above)
Comment (arajasek) [reply actors/migration/nv15/miner.go:64]: Yessir!

File: actors/migration/nv15/miner.go (hunk 6 above)
Comment (arajasek) [reply actors/migration/nv15/miner.go:72]: sg, i think it's better to error -- doing something silently wrong is probably worse

File: actors/migration/nv15/miner.go [hunk 11]
```diff
@@ -3,79 +3,201 @@ package nv15
...
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to construct new sectors array: %w", err)
+			}
+
+			var sectorInfo miner6.SectorOnChainInfo
+			if err = inArray.ForEach(&sectorInfo, func(k int64) error {
+				return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
+			}); err != nil {
+				return cid.Undef, err
+			}
```
Comment (Stebalien) [inline actors/migration/nv15/miner.go:92]: Followup: https://github.com/filecoin-project/go-amt-ipld/issues/61

File: actors/migration/nv15/miner.go [hunk 12]
```diff
@@ -3,79 +3,201 @@ package nv15
...
+		return cid.Undef, xerrors.Errorf("unexpected WPoStPeriodDeadlines changed from %d to %d",
+			miner6.WPoStPeriodDeadlines, miner7.WPoStPeriodDeadlines)
 	}
 
-	var sectorInfo miner6.SectorOnChainInfo
-	if err = inArray.ForEach(&sectorInfo, func(k int64) error {
-		return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
-	}); err != nil {
-		return cid.Undef, err
+	outDeadlines := miner7.Deadlines{Due: [miner7.WPoStPeriodDeadlines]cid.Cid{}}
```
Comment (Stebalien) [inline actors/migration/nv15/miner.go:167]: Does this not work?

```suggestion
	var outDeadlines miner7.Deadlines
```

File: actors/migration/nv15/miner.go [hunk 13]
```diff
@@ -3,79 +3,201 @@ package nv15
...
+
+	for i, c := range inDeadlines.Due {
+		outDlCid, err := cache.Load(DeadlineKey(c), func() (cid.Cid, error) {
```
Comment (Stebalien) [inline actors/migration/nv15/miner.go:170]: This assumption is wrong. If the sectors CID has changed, we need to migrate the deadlines regardless.

We have two options:

1. Just migrate every time. Honestly, this is probably _fine_ as long as we're caching top-level actor migrations. If a miner is _active_, their deadlines will have changed significantly anyways. If a miner isn't active, the top-level actor cache will skip the miner.
2. Key by deadline CID _and_ sectors CID.

File: actors/migration/nv15/miner.go [hunk 14]
```diff
@@ -3,79 +3,201 @@ package nv15
...
+		}
+
+		okOut, prevOutRoot, err := cache.Read(MinerPrevSectorsOutKey(minerAddr))
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to get previous outRoot from cache: %w", err)
+		}
+
+		if okIn != okOut {
+			return cid.Undef, xerrors.Errorf("cannot have only one of the inRoot and outRoot in the cache")
+		}
```
Comment (Stebalien) [inline actors/migration/nv15/miner.go:77]: If we get an error here, we may want to just log it, assume _nothing_ is cached, and do the migration.

File: actors/migration/nv15/miner.go [hunk 15]
```diff
@@ -3,79 +3,201 @@ package nv15
...
+			return cid.Undef, xerrors.Errorf("error writing new sectors AMT: %w", err)
+		}
+
+		if err = cache.Write(MinerPrevSectorsInKey(minerAddr), inRoot); err != nil {
+			return cid.Undef, xerrors.Errorf("failed to write previous sectors in key to cache: %w", err)
+		}
+
+		if err = cache.Write(MinerPrevSectorsOutKey(minerAddr), outRoot); err != nil {
+			return cid.Undef, xerrors.Errorf("failed to write previous sectors out key to cache: %w", err)
+		}
```
Comment (Stebalien) [inline actors/migration/nv15/miner.go:150]: Again, consider logging an error and moving on.

File: actors/migration/nv15/top.go [hunk 16]
```diff
@@ -57,6 +53,28 @@ type MigrationCache interface {
...
+	return addr.String() + "-h-" + head.String()
+}
+
+func SectorsAmtKey(sectorsAmt cid.Cid) string {
+	return "s-" + sectorsAmt.String()
+}
+
+func DeadlineKey(dl cid.Cid) string {
+	return "d-" + dl.String()
+}
```
Comment (Stebalien) [inline actors/migration/nv15/top.go:66]: Can we use `StringOfBase` here? Just to be clear.

File: actors/migration/nv15/top.go [hunk 17]
```diff
@@ -57,6 +53,28 @@ type MigrationCache interface {
...
+
+func MinerPrevSectorsInKey(m address.Address) string {
+	// "i" for "in"
+	return "i-" + m.String()
```
Comment (Stebalien) [inline actors/migration/nv15/top.go:70]: nit: we can afford a few more characters.

File: actors/migration/nv15/top.go (hunk 3 above)
Comment (Stebalien) [reply actors/migration/nv15/top.go:249]: I think we needed it for some migrations in the past.

**Review by Stebalien — CHANGES_REQUESTED** [review]

File: actors/migration/nv15/miner.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (arajasek) [reply actors/migration/nv15/miner.go:77]: I was lazy to do so because the cache as currently written can only have a CID as the `value` type :P

File: actors/migration/nv15/miner.go (hunk 12 above)
Comment (arajasek) [reply actors/migration/nv15/miner.go:167]: It does, I just copied the v3 migration

File: actors/migration/nv15/miner.go (hunk 13 above)
Comment (arajasek) [reply actors/migration/nv15/miner.go:170]: Yessir, you right -- I think I'm gonna migrate every time, with a special case for the empty deadline and deadlines objects which I think will be a sizeable difference.

File: actors/migration/nv15/miner.go (hunk 14 above)
Comment (arajasek) [reply actors/migration/nv15/miner.go:77]: Yeah, this was a convo between Zen and me upstream^, neither of us felt too strongly...I think I'll adopt your suggestion

File: actors/migration/nv15/top.go (hunk 16 above)
Comment (arajasek) [reply actors/migration/nv15/top.go:66]: Looks like that would mean that the Key methods would have to return an error (or do something else about the error), which isn't great...

Comment (Stebalien) [reply actors/migration/nv15/top.go]: Panic? That can only happen if the encoding isn't supported.

Basically, `dl.String()` isn't guaranteed to a _specific_ base.

File: actors/migration/nv15/miner.go [hunk 18]
```diff
@@ -3,79 +3,251 @@ package nv15
...
+				return cid.Undef, err
+			}
+		}
+
+		outRoot, err := outArray.Root()
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("error writing new sectors AMT: %w", err)
+		}
+
+		_ = cache.Write(MinerPrevSectorsInKey(minerAddr), inRoot)
```
Comment (arajasek) [inline actors/migration/nv15/miner.go:199]: Silently dropping the error here, because I wasn't sure whether concurrent access of the logger would be safe...anyone know?

File: actors/migration/nv15/top.go (hunk 16 above)
Comment (arajasek) [reply actors/migration/nv15/top.go]: Sure, I was too scared to do that :P

**Comment by arajasek** [comment]
Rebased on top of latest master (revert of deal label PR)

**Comment by arajasek** [comment]
Testing update: Ran on mainnet state with and without the cache, same result. 

Next steps are a unit test coming from @geoff-vball and maybe writing an unmigration, and testing that.

File: actors/migration/nv15/miner.go (hunk 18 above)
Comment (Stebalien) [reply actors/migration/nv15/miner.go:199]: Yes: https://github.com/filecoin-project/lotus/blob/dc26f3bd7453c515ca9aecc16fb9b3517784ef5e/chain/consensus/filcns/upgrades.go#L1205

File: actors/migration/nv15/miner.go [hunk 19]
```diff
@@ -3,79 +3,251 @@ package nv15
...
+
+		_ = cache.Write(MinerPrevSectorsOutKey(minerAddr), outRoot)
+		return outRoot, nil
+	})
+}
+
+func (m minerMigrator) migrateDeadlines(ctx context.Context, store adt.Store, deadlines cid.Cid, sectors cid.Cid) (cid.Cid, error) {
+	if deadlines == m.emptyDeadlinesV6 && sectors == m.emptySectorsV7 {
```
Comment (Stebalien) [inline actors/migration/nv15/miner.go:207]: Why do I care if the sectors are empty?

File: actors/migration/nv15/miner.go [hunk 20]
```diff
@@ -3,79 +3,251 @@ package nv15
...
 	}
 
-	var sectorInfo miner6.SectorOnChainInfo
-	if err = inArray.ForEach(&sectorInfo, func(k int64) error {
-		return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
-	}); err != nil {
-		return cid.Undef, err
+	var outDeadlines miner7.Deadlines
+	for i, c := range inDeadlines.Due {
+		if c == m.emptyDeadlineV6 && sectors == m.emptySectorsV7 {
```
Comment (Stebalien) [inline actors/migration/nv15/miner.go:219]: I'm not sure if this is correct. Usually, when we hit a "dead" deadline, we skip it and won't record any snapshots: https://github.com/filecoin-project/specs-actors/blob/d46a6510c2ad20ca453f72e292adbe880a8bacb8/actors/builtin/miner/miner_state.go#L1149. That means that never-used deadlines shouldn't have sector snapshots.

This is actually a really interesting edge-case because we can run into a state where we keep a _really old_ copy of the sectors array and never update it (if a deadline "dies"). This isn't a problem for _partitions_ because we only snapshot the deadline's partitions (so, when the deadline "dies", the partitions array and the partitions snapshot array are identical).

File: actors/migration/nv15/miner.go (hunk 19 above)
Comment (arajasek) [reply actors/migration/nv15/miner.go:207]: Just being extrachecky -- i'm not really sure what it means to have an empty deadlines but non empty sectors, so i decided to say migrate it manually if that happens :P

**Review by Stebalien — APPROVED** [review]
We're going to address my comments in a new PR as we need to change behavior a bit.

**Comment by arajasek** [comment]
Landing this, thanks Steb!
